main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -0,0 +1,235 @@
+        // {}
+        case (union@Union(_, _, _, _, _, _, _), Empty) =>
+          p.progress()
+          val eqs = union.mapSubformulas(Equation.mk(_, Empty, loc))
+          (eqs, SetSubstitution.empty)
+
+        // f1 ~ f2, where f1 and f2 are ground
+        // ---
+        // [], {} if solved
+        // [f1 !~ f2], {} if unsolvable
+        case (f1, f2) if f1.isGround && f2.isGround =>
+          p.progress()
+          if (isEquivalent(f1, f2)) (Nil, SetSubstitution.empty)
+          else (List(eq.toUnsolvable), SetSubstitution.empty)
+
+        case _ =>
+          // Can't do anything.
+          (List(eq), SetSubstitution.empty)
+      }
+    }
+
+    /**
+      * Solves equations variable alias equations (e.g. `x1 ~ x2`).
+      *
+      *   - `x1 ~ x2` becomes `({}, [x1 -> x2])`
+      *   - `!x1 ~ !x2` becomes `({}, [x1 -> x2])`
+      */
+    def variableAlias(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      val Equation(t1, t2, _, _) = eq
+      (t1, t2) match {
+        // x1 ~ x2
+        // ---
+        // {},
+        // [x1 -> x2]
+        case (Var(x), y@Var(_)) =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, y))
+
+        // !x1 ~ !x2
+        // ---
+        // {},
+        // [x1 -> x2]
+        case (Compl(Var(x)), Compl(y@Var(_))) =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, y))
+
+        case _ =>
+          // Can't do anything.
+          (List(eq), SetSubstitution.empty)
+      }
+    }
+
+    /**
+      * Solves non-recursive variable assignments (e.g. `x1 ~ x2 ∪ c4`).
+      *
+      *   - `x ~ f` where `f` does not contain `x` becomes `([], {x -> f})`
+      *   - `!x ~ f` where `f` does not contain `x` becomes `([], {x -> !f})`
+      */
+    def variableAssignment(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      val Equation(t1, t2, _, _) = eq
+      (t1, t2) match {
+        // x ~ f, where f does not contain x
+        // ---
+        // {},
+        // [x -> f]
+        case (v@Var(x), f) if !f.contains(v) =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, f))
+
+        // !x ~ f, where f does not contain x
+        // ---
+        // {},
+        // [x -> !f]
+        case (Compl(v@Var(x)), f) if !f.contains(v) =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, mkCompl(f)))
+
+        case _ =>
+          // Can't do anything.
+          (List(eq), SetSubstitution.empty)
+      }
+    }
+
+    /**
+      * Solves equations using successive-variable-elimination, i.e. exhaustive instantiation.
+      *
+      * Always returns no equations or `eq` marked as [[Equation.Status.Unsolvable]] or
+      * [[Equation.Status.Timeout]].
+      */
+    def sve(recSizeThreshold: Int)(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      p.progress()
+      val query = mkEquivalenceTestToEmpty(eq.f1, eq.f2)
```

Comment:
mkEquivalenceTestToEmpty - I find this name hard to understand.

Reply:
yea i found it hard to name - I could just call it `mkXor` but that also doesn't why that is what we want.

`mkEmptyTester` ?

Reply:
Oh no, not mkEmptyTester

Reply:
mkEmptyQuery?

Reply:
Yes